Feature/sso

feedmypixel opened the pull request:
# Add support for Single sign-on (SSO)

This work is paired  and confirmed to be working with [#598](https://github.com/uktrade/data-hub-leeloo/pull/598)

## Details
This work involves
- [x] SSO journey
- [x] SSOBypass work for developers
- [x] updates to README
- [x] tests
- [x] Getting SSO working on CircleCI
### Of note
UAT tests will fail because I need to setup OAuth on CircleCi (tomorrow)

 The journey is:
- if not already logged in
- redirects to OAuth
- which then sends information to our callback url
- which then takes this information and gets the Access token
- which is then used across the app

This gif shows me logging in via OAuth.
![sso](https://user-images.githubusercontent.com/2305016/32349100-e3f0c248-c00d-11e7-8145-2c90e30200ee.gif)


### Minimal Help page 

![screen shot 2017-11-04 at 12 51 03](https://user-images.githubusercontent.com/2305016/32405579-112b494c-c160-11e7-8d36-a843ed22eab5.png)

ztolley commented:
If we are using oAuth2, could we have used Passport?

ztolley commented:
Nice!

ztolley commented on src/apps/oauth/controllers.js:
```diff
@@ -0,0 +1,64 @@
+const uuid = require('uuid')
+const { set } = require('lodash')
+
+const config = require('./../../../config')
```
doesn't need the first `./`

feedmypixel replied:
I prefer this for clarities sake

yeahfro commented on test/unit/apps/oauth/controllers/oauth.test.js:
```diff
@@ -0,0 +1,178 @@
+  })
+
+  describe('#oauthCallback', () => {
+    context('callback state query param', () => {
```
I feel this would read better `when there is a state query param` (just going by the unit test visualisation in WebStorm)

yeahfro commented on test/unit/apps/oauth/controllers/oauth.test.js:
```diff
@@ -0,0 +1,178 @@
+      })
+    })
+
+    context('callback error query param', () => {
```
I feel this would read better `when there is a query param`

yeahfro commented on test/unit/apps/oauth/controllers/oauth.test.js:
```diff
@@ -0,0 +1,178 @@
+      })
+    })
+
+    context('#getBearerToken', () => {
```
is this `context` right? It doesn't really make much sense to me

yeahfro commented on test/unit/apps/oauth/controllers/oauth.test.js:
```diff
@@ -0,0 +1,178 @@
+  describe('#oauthCallback', () => {
+    context('callback state query param', () => {
+      it('should show error with stateId mismatch', () => {
+        set(this.reqMock, 'query.state', 'mock-state-id')
```
I can see `mock-state-id` being set as a variable in the next test. Can anything be done to keep these consistent?

yeahfro commented on test/unit/apps/oauth/controllers/oauth.test.js:
```diff
@@ -0,0 +1,178 @@
+    context('callback state query param', () => {
+      it('should show error with stateId mismatch', () => {
+        set(this.reqMock, 'query.state', 'mock-state-id')
+        set(this.reqMock, 'session.oauth.stateId', 'non-matching-state-id')
```
Really into this use of `set`

feedmypixel commented:
> I assume there are more pr's to come that will have a more user-friendly error flow?

Yeah some work is being done with Live Services around this.

feedmypixel commented:
> I assume there are more pr's to come that will have a more user-friendly error flow?

FYI I have just added a minimal page for this and updated tests accordingly

yeahfro commented on test/acceptance/features/events/create.feature:
```diff
@@ -4,67 +4,64 @@ Feature: Create an Event in Data hub
-    And I verify the shared teams field is not displayed
-    And I verify the event related programmes field is displayed
-    And I verify the event save button is displayed
+    Then I verify the event type field is displayed
```
How come all of these have ended up as `Then`?
